### ISSUE TITLE
Hard refresh speed

### ISSUE THREAD

**Issue by natew** [steelbrain/pundle#140]
Not to bug you too much as this is a pretty small one, but figured I'd note. I know we looked at this previously too. When I refresh our page I see "waiting for localhost" for what seems like a few seconds (2-3) before it loads. Wondering if it's doing any extra work that couldn't be cached.

Also, lovely to see you making progress! 🎉

**Comment by steelbrain** [comment]
To make HMR fast, we have to defer the regeneration of the full bundle until when it's requested. To make hard refresh faster, we'll have to sacrifice HMR speed. Furthermore if you hard refresh multiple times without changing the bundle, the regeneration will happen only once.

We could probably speed it up in other places to make hard refresh regeneration look faster but it'll definitely take the time it does if you have a big bundle

**Comment by natew** [comment]
Ah this makes sense, thanks for explanation. Perhaps some sort of throttled watch to rebuilds the full thing at most every 1s to prevent too much overwork. Definitely in "nice-to-have" enhancement category.

**Comment by steelbrain** [comment]
In a realistic scenario, you would HMR say 10-20 times, maybe even more at intervals more than 1s, so we would be generating those 10-20 times unnecessarily and if you are doing it on a laptop, the fan would go crazy.

I think it's good to have it recompile on pull, reduces any redundant regenerations and 1-2s isn't much if you don't do it too often (which you shouldn't because you have HMR)

**Comment by natew** [comment]
Still think theres something going on here. This is time spend after relading 4-5 times without making any changes:

![image](https://cloud.githubusercontent.com/assets/12100/26765244/a99f100c-492c-11e7-8226-40c93d2c3952.png)

Notice "Waiting" is 6s. Thats about right for how long it takes generally. Actually this may be low hanging enough I'd be able to get a PR through here helping. My main concern is I'm on a slightly older branch while waiting for master to work. Any hope? I'm going to poke through code today but tips appreciated.

**Comment by steelbrain** [comment]
If you could explain how you're going to solve this before opening a PR, that would be perfect as it would give me insight on how I could've improved it myself

**Comment by natew** [comment]
I spent some time reading the src and got at least familiar with high level what is happening.

What I'm saying here is that it seems to be doing a lot of work every refresh, even without changes.

To me the fix would be really simple:

- Watch for changes
  - on change mark dirty
- On reload
  - if dirty, rebuild and cache result
  - if not dirty serve cached result

Basically, a really simple watch for if its dirty, and if not serve the cached bundle. I assume it's doing some sort of extra work by re-building the bundle on every refresh.

**Comment by steelbrain** [comment]
We can definitely fix the part where it regenerates without changes. The change SGTM, would love to review a PR doing that

**Comment by natew** [comment]
Just writing as I think this is one of the few things that would make a big diff for a v2, this + bugfixes + docs would be perfect and I can help with docs.

**Comment by steelbrain** [comment]
Hard refresh speed is now incredibly fast in v3 alpha as files are already generated and just need to be chunk-generated (aka concatenated before being sent to the client) so this is no longer an issue
